src/main/java/org/opensearch/geospatial/ip2geo/action/PutDatasourceTransportAction.java

```diff
@@ -91,99 +100,106 @@ public PutDatasourceTransportAction(
+        } catch (Exception e) {
+            log.error("Failed to create datasource for {}", datasource.getId(), e);
+            markDatasourceAsCreateFailed(datasource);
+        }
+    }
+
+    private void markDatasourceAsCreateFailed(final Datasource datasource) {
+        datasource.getUpdateStats().setLastFailedAt(Instant.now());
+        datasource.setState(DatasourceState.CREATE_FAILED);
+        try {
+            datasourceHelper.updateDatasource(datasource);
+        } catch (Exception e) {
+            log.error("Failed to mark datasource state as CREATE_FAILED for {}", datasource.getId(), e);
+        }
     }
 
     private void updateJobParameterAsSucceeded(
-        final Datasource jobParameter,
+        final Datasource datasource,
         final DatasourceManifest manifest,
-        final String[] fields,
+        final List<String> fields,
         final Instant startTime,
         final Instant endTime
     ) throws IOException {
-        jobParameter.setDatabase(manifest, fields);
-        jobParameter.getUpdateStats().setLastSucceededAt(endTime);
-        jobParameter.getUpdateStats().setLastProcessingTimeInMillis(endTime.toEpochMilli() - startTime.toEpochMilli());
-        jobParameter.enable();
-        jobParameter.setState(DatasourceState.AVAILABLE);
-        DatasourceHelper.updateDatasource(client, jobParameter, timeout);
+        datasource.setDatabase(manifest, fields);
+        datasource.getUpdateStats().setLastSucceededAt(endTime);
+        datasource.getUpdateStats().setLastProcessingTimeInMillis(endTime.toEpochMilli() - startTime.toEpochMilli());
+        datasource.enable();
+        datasource.setState(DatasourceState.AVAILABLE);
+        datasourceHelper.updateDatasource(datasource);
     }
 
-    private String setupIndex(final DatasourceManifest manifest, final Datasource jobParameter) throws IOException {
-        String indexName = jobParameter.indexNameFor(manifest);
-        jobParameter.getIndices().add(indexName);
-        DatasourceHelper.updateDatasource(client, jobParameter, timeout);
-        GeoIpDataHelper.createIndexIfNotExists(clusterService, client, indexName, timeout);
+    private String setupIndex(final DatasourceManifest manifest, final Datasource datasource) throws IOException {
+        String indexName = datasource.indexNameFor(manifest);
+        datasource.getIndices().add(indexName);
+        datasourceHelper.updateDatasource(datasource);
+        geoIpDataHelper.createIndexIfNotExists(indexName);
         return indexName;
     }
 
-    private String[] putIp2GeoData(final String indexName, final DatasourceManifest manifest) throws IOException {
-        String[] fields;
-        try (CSVParser reader = GeoIpDataHelper.getDatabaseReader(manifest)) {
+    private List<String> putGeoIpData(final String indexName, final DatasourceManifest manifest) throws IOException {
```

Comment:
put method returning header seems odd. Shall we split it into two method? put and getHeader for better readability?

Reply:
Removed the method.